From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Reilly Brogan <reilly@reillybrogan.com>
Date: Wed, 12 Jun 2024 15:24:53 -0500
Subject: [PATCH] downstream: Use kirigami mousewheel to fix scrolling

The default Qt6 ListView mouse wheel scrolling feels terrible, use the kirigami WheelHandler to make that feel "right"

Signed-off-by: Reilly Brogan <reilly@reillybrogan.com>
---
 CMakeLists.txt                                       | 10 ++++++++++
 resources/qml/CommunitiesList.qml                    |  8 ++++++++
 resources/qml/MessageView.qml                        |  8 ++++++++
 resources/qml/RoomList.qml                           |  8 ++++++++
 resources/qml/dialogs/AllowedRoomsSettingsDialog.qml |  7 +++++++
 resources/qml/dialogs/IgnoredUsers.qml               |  8 ++++++++
 resources/qml/dialogs/RoomDirectory.qml              |  8 ++++++++
 resources/qml/dialogs/RoomMembers.qml                |  8 ++++++++
 resources/qml/dialogs/UserProfile.qml                |  8 ++++++++
 9 files changed, 73 insertions(+)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index c513bace..bf082aa9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -270,6 +270,15 @@ else()
     find_package(KDSingleApplication-qt6 REQUIRED)
 endif()
 
+set(KF_MIN_VERSION "6.0.0")
+find_package(ECM ${KF_MIN_VERSION} REQUIRED NO_MODULE)
+list(APPEND CMAKE_MODULE_PATH ${ECM_MODULE_PATH})
+include(ECMFindQmlModule)
+find_package(KF6 REQUIRED COMPONENTS
+    Kirigami
+)
+ecm_find_qmlmodule(org.kde.kirigami REQUIRED)
+
 if(Qt6Widgets_FOUND)
     if(Qt6Widgets_VERSION VERSION_LESS 6.5.0)
         message(STATUS "Qt version ${Qt6Widgets_VERSION}")
@@ -904,6 +913,7 @@ target_link_libraries(nheko PRIVATE
     Qt::Qml
     Qt::QmlPrivate
     Qt::QuickControls2
+    KF6::Kirigami
     qt6keychain
     KDAB::kdsingleapplication
     nlohmann_json::nlohmann_json
diff --git a/resources/qml/CommunitiesList.qml b/resources/qml/CommunitiesList.qml
index c9f259e2..630a020c 100644
--- a/resources/qml/CommunitiesList.qml
+++ b/resources/qml/CommunitiesList.qml
@@ -9,6 +9,7 @@ import QtQuick
 import QtQuick.Controls
 import QtQuick.Layouts
 import im.nheko
+import org.kde.kirigami as Kirigami
 
 Page {
     id: communitySidebar
@@ -39,6 +40,13 @@ Page {
         model: Communities.filtered()
         boundsBehavior: Flickable.StopAtBounds
 
+        Kirigami.WheelHandler {
+            id: wheelHandler
+            target: communitiesList
+            filterMouseEvents: true
+            keyNavigationEnabled: true
+        }
+
         ScrollBar.vertical: ScrollBar {
             id: scrollbar
 
diff --git a/resources/qml/MessageView.qml b/resources/qml/MessageView.qml
index f253b7a8..2447caf2 100644
--- a/resources/qml/MessageView.qml
+++ b/resources/qml/MessageView.qml
@@ -10,6 +10,7 @@ import QtQuick.Controls 2.15
 import QtQuick.Layouts 1.2
 import QtQuick.Window 2.13
 import im.nheko 1.0
+import org.kde.kirigami as Kirigami
 
 Item {
     id: chatRoot
@@ -66,6 +67,13 @@ Item {
         spacing: 2
         verticalLayoutDirection: ListView.BottomToTop
 
+        Kirigami.WheelHandler {
+            id: wheelHandler
+            target: chat
+            filterMouseEvents: true
+            keyNavigationEnabled: true
+        }
+
         property int lastScrollPos: 0
 
         // Fixup the scroll position when the height changes. Without this, the view is kept around the center of the currently visible content, while we usually want to stick to the bottom.
diff --git a/resources/qml/RoomList.qml b/resources/qml/RoomList.qml
index 350b3846..a3a6cd60 100644
--- a/resources/qml/RoomList.qml
+++ b/resources/qml/RoomList.qml
@@ -11,6 +11,7 @@ import QtQuick
 import QtQuick.Controls
 import QtQuick.Layouts
 import im.nheko
+import org.kde.kirigami as Kirigami
 
 Page {
     //leftPadding: Nheko.paddingSmall
@@ -440,6 +441,13 @@ Page {
         model: Rooms
         boundsBehavior: Flickable.StopAtBounds
 
+        Kirigami.WheelHandler {
+            id: wheelHandler
+            target: roomlist
+            filterMouseEvents: true
+            keyNavigationEnabled: true
+        }
+
         //reuseItems: true
         ScrollBar.vertical: ScrollBar {
             id: scrollbar
diff --git a/resources/qml/dialogs/AllowedRoomsSettingsDialog.qml b/resources/qml/dialogs/AllowedRoomsSettingsDialog.qml
index 82b5506e..1d1ad108 100644
--- a/resources/qml/dialogs/AllowedRoomsSettingsDialog.qml
+++ b/resources/qml/dialogs/AllowedRoomsSettingsDialog.qml
@@ -8,6 +8,7 @@ import QtQuick.Controls
 import QtQuick.Layouts
 import QtQuick.Window
 import im.nheko
+import org.kde.kirigami as Kirigami
 
 ApplicationWindow {
     id: allowedDialog
@@ -51,6 +52,12 @@ ApplicationWindow {
 
             clip: true
 
+            Kirigami.WheelHandler {
+                id: wheelHandler
+                target: view
+                filterMouseEvents: true
+                keyNavigationEnabled: true
+            }
 
             model: roomSettings.allowedRoomsModel
             spacing: 4
diff --git a/resources/qml/dialogs/IgnoredUsers.qml b/resources/qml/dialogs/IgnoredUsers.qml
index 6d6585f0..54eac9c6 100644
--- a/resources/qml/dialogs/IgnoredUsers.qml
+++ b/resources/qml/dialogs/IgnoredUsers.qml
@@ -8,6 +8,7 @@ import QtQuick.Controls
 import QtQuick.Layouts
 import QtQuick.Window
 import im.nheko
+import org.kde.kirigami as Kirigami
 import "../"
 
 Window {
@@ -26,6 +27,13 @@ Window {
         spacing: Nheko.paddingMedium
         footerPositioning: ListView.OverlayFooter
 
+        Kirigami.WheelHandler {
+            id: wheelHandler
+            target: view
+            filterMouseEvents: true
+            keyNavigationEnabled: true
+        }
+
         model: TimelineManager.ignoredUsers
         header: ColumnLayout {
             Text {
diff --git a/resources/qml/dialogs/RoomDirectory.qml b/resources/qml/dialogs/RoomDirectory.qml
index dbf3b459..a50b6926 100644
--- a/resources/qml/dialogs/RoomDirectory.qml
+++ b/resources/qml/dialogs/RoomDirectory.qml
@@ -9,6 +9,7 @@ import QtQuick.Controls 2.3
 import QtQuick.Layouts 1.3
 import QtQuick.Window 2.15
 import im.nheko 1.0
+import org.kde.kirigami as Kirigami
 
 ApplicationWindow {
     id: roomDirectoryWindow
@@ -34,6 +35,13 @@ ApplicationWindow {
         anchors.fill: parent
         model: publicRooms
 
+        Kirigami.WheelHandler {
+            id: wheelHandler
+            target: roomDirView
+            filterMouseEvents: true
+            keyNavigationEnabled: true
+        }
+
         delegate: Rectangle {
             id: roomDirDelegate
 
diff --git a/resources/qml/dialogs/RoomMembers.qml b/resources/qml/dialogs/RoomMembers.qml
index 95dc9fc3..69f182a3 100644
--- a/resources/qml/dialogs/RoomMembers.qml
+++ b/resources/qml/dialogs/RoomMembers.qml
@@ -10,6 +10,7 @@ import QtQuick.Controls 2.12
 import QtQuick.Layouts 1.12
 import QtQuick.Window 2.13
 import im.nheko 1.0
+import org.kde.kirigami as Kirigami
 
 ApplicationWindow {
     id: roomMembersRoot
@@ -108,6 +109,13 @@ ApplicationWindow {
                 boundsBehavior: Flickable.StopAtBounds
                 model: members
 
+                Kirigami.WheelHandler {
+                    id: wheelHandler
+                    target: memberList
+                    filterMouseEvents: true
+                    keyNavigationEnabled: true
+                }
+
 
                 delegate: ItemDelegate {
                     id: del
diff --git a/resources/qml/dialogs/UserProfile.qml b/resources/qml/dialogs/UserProfile.qml
index 6798bc79..532155bc 100644
--- a/resources/qml/dialogs/UserProfile.qml
+++ b/resources/qml/dialogs/UserProfile.qml
@@ -11,6 +11,7 @@ import QtQuick.Layouts 1.2
 import QtQuick.Window 2.13
 import QtQml.Models 2.2
 import im.nheko 1.0
+import org.kde.kirigami as Kirigami
 
 ApplicationWindow {
     id: userProfileDialog
@@ -45,6 +46,13 @@ ApplicationWindow {
         anchors.margins: 10
         footerPositioning: ListView.OverlayFooter
 
+        Kirigami.WheelHandler {
+            id: wheelHandler
+            target: devicelist
+            filterMouseEvents: true
+            keyNavigationEnabled: true
+        }
+
         header: ColumnLayout {
             id: contentL
 
